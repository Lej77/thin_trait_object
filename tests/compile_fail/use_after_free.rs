use thin_trait_object::prelude::*;

define_v_table!(
    trait ToStringVTable {
        fn dyn_to_string(&self) -> String;
    }
);

struct Wrapper<T: ?Sized>(T);

impl<T> ToStringVTable for Wrapper<T>
where
    T: ToString + ?Sized,
{
    fn dyn_to_string(&self) -> String {
        self.0.to_string()
    }
}

fn main() {
    {
        let text = "test".to_owned();
        // Works:
        ThinBox::<'_, dyn ToStringVTable, _>::new(Wrapper(&text), ());
        // Fails:
        ThinBox::<'static, dyn ToStringVTable, _>::new(Wrapper(&text), ());
    };

    let erased = {
        let text = "test".to_owned();
        // Fails:
        ThinBox::<'_, dyn ToStringVTable, _>::new(Wrapper(&text), ())
    };
    assert_eq!(erased.dyn_to_string(), "test");

    let erased = {
        let text = "test".to_owned();
        // Fails:
        ThinBoxWithoutCommon::<'_, dyn ToStringVTable, _>::new(Wrapper(&text))
    };
    assert_eq!(erased.dyn_to_string(), "test");
}
