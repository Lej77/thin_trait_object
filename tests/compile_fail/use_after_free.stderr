error[E0597]: `text` does not live long enough
  --> $DIR/use_after_free.rs:26:64
   |
26 |         ThinBox::<'static, dyn ToStringVTable, _>::new(Wrapper(&text), ());
   |         -------------------------------------------------------^^^^^------
   |         |                                                      |
   |         |                                                      borrowed value does not live long enough
   |         argument requires that `text` is borrowed for `'static`
27 |     };
   |     - `text` dropped here while still borrowed

error[E0597]: `text` does not live long enough
  --> $DIR/use_after_free.rs:32:59
   |
29 |     let erased = {
   |         ------ borrow later stored here
...
32 |         ThinBox::<'_, dyn ToStringVTable, _>::new(Wrapper(&text), ())
   |                                                           ^^^^^ borrowed value does not live long enough
33 |     };
   |     - `text` dropped here while still borrowed

error[E0597]: `text` does not live long enough
  --> $DIR/use_after_free.rs:39:72
   |
36 |     let erased = {
   |         ------ borrow later stored here
...
39 |         ThinBoxWithoutCommon::<'_, dyn ToStringVTable, _>::new(Wrapper(&text))
   |                                                                        ^^^^^ borrowed value does not live long enough
40 |     };
   |     - `text` dropped here while still borrowed
